### ISSUE TITLE
qloss

### ISSUE THREAD

**Issue by hbellafkir** [Minione/DTSH#2]
hi,

how its possible that you're using sign function for the quantization loss  when the sign function is not differentiable?  that doesn't make since, when the majority of papers avoids the sign function and uses tanh activation as approximation. 

thanks in advance

**Comment by Minione** [comment]
I guess you meant L40 in train_triplet.m. qloss is the quantization error term. Here we try to encourage the learned code to be binary (-1 or 1). We use sign(U0) to represent the binary codes after quantization/binarization.

Mathematically qloss = |U0-B0|_2, so the gradient of qloss is well defined and is 2(U0-B0), where B0=sign(U0).

**Comment by hbellafkir** [comment]
How? you have to derive sgn-function too (chain rule).  The derivation is respect to U0.

**Comment by Minione** [comment]
The sign function sign(x) is differentiable except one point x=0, but the gradient of sign function is 0. As said above, the gradient of qloss w.r.t. U0 is 2(U0-B0) and that's exactly the result of the chain rule.

**Comment by hbellafkir** [comment]
Ok I got it thanks. For the others wondering:
d (|U0- sgn(U0)|_2) /dU0 =  2(U0-sing(U0)) *  d (U0-sgn(U0)) / dU0 = 2(U0-sing(U0))
